scripts/upower/95-upower-hid.rules

```diff
@@ -1,192 +1,2 @@
-##############################################################################################################
-# Uninterruptible Power Supplies with USB HID interfaces
-#
-# This file was automatically generated by NUT:
-# https://github.com/networkupstools/nut/
-#
-# To keep up to date, monitor upstream NUT
-# https://github.com/networkupstools/nut/commits/master/scripts/upower/95-upower-hid.rules
-# or checkout the NUT repository and call 'tools/nut-usbinfo.pl'
-
-# newer hiddev are part of the usbmisc class
-SUBSYSTEM=="usbmisc", GOTO="up_hid_chkdev"
-# only support USB, else ignore
-SUBSYSTEM!="usb", GOTO="up_hid_end"
-
-# if usbraw device, ignore
-LABEL="up_hid_chkdev"
-KERNEL!="hiddev*", GOTO="up_hid_end"
-
-# if an interface, ignore
-ENV{DEVTYPE}=="usb_interface", GOTO="up_hid_end"
```

Comment:
from udev(7)
```
       SUBSYSTEM
           Match the subsystem of the event device.

       SUBSYSTEMS
           Search the devpath upwards for a matching device subsystem name.     
```
 * The hwdb tags USB devices (not the usbmisc device!)
 * We then match a "usbmisc" device that has a "usb" parent (which is probably redundant), so yes, it is honoured
 * then, with KERNEL=="hiddev*", exactly the same match as before, just inverted for obvious reasons
 * then, with IMPORT{}, we copy those tags down to the hiddev

It is not 100% equivalent. As, it for example copies the ID_PRODUCT/ID_VENDOR entries down to the `usbmisc` device (not sure why I did that, that might actually be redundant and `IMPORT{UPOWER_*}` should be sufficient).

I think `usb_interface` is just an optimization. i.e. don't bother with executing hundreds of rules for *every* endpoint. So yes, the exemption is missing but it will not matter as it is only one simple match/import rule. But, we can also add it back without any negative side effect. After all, no one needs UPOWER_* variables on every endpoint.

Reply:
Another difference is obviously that the USB device itself also gets the UPOWER_* variables, which before it had not.

Reply:
My reading was that this original matching was a sort of "if/switch" statement:
* if `SUBSYSTEM==usbmisc`, go to mid-way label `up_hid_chkdev`
* if `SUBSYSTEM==usb`, go to mid-way label `up_hid_chkdev` (by default, because otherwise we bail)

So if `SUBSYSTEM` is any one of these two - try next steps (consider hiddev and interfaces); otherwise don't bother.

Reply:
As for interfaces and settings on "every endpoint", not being a domain expert I am in fact not sure what we should want here technically. In the past couple of years more and more devices appear which serve UPS HID protocols on a non-zero USB interface number, through libusb. If these translate on Linux to different filesystem nodes with their separate access permissions, and if this is the filter which separates base device (and/or interface #0 used for ages) from the non-zero interfaces, then perhaps we do want to skip the filter. It might however break some other software, I guess (whose generic/storage/scanner/whatever endpoints would get owned by `nut` and inaccessible to their specific software).

I just don't know for sure here, so voicing what looks like reasonable-guess concerns.